sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala

```diff
@@ -197,34 +199,50 @@ class UnivocityParser(
         Decimal(decimalParser(datum), dt.precision, dt.scale)
       }
 
-    case _: TimestampType => (d: String) =>
+    case _: DateType => (d: String) =>
       nullSafeDatum(d, name, nullable, options) { datum =>
         try {
-          timestampFormatter.parse(datum)
+          dateFormatter.parse(datum)
         } catch {
           case NonFatal(e) =>
             // If fails to parse, then tries the way used in 2.0 and 1.x for backwards
             // compatibility.
             val str = DateTimeUtils.cleanLegacyTimestampStr(UTF8String.fromString(datum))
-            DateTimeUtils.stringToTimestamp(str, options.zoneId).getOrElse(throw e)
+            DateTimeUtils.stringToDate(str).getOrElse(throw e)
         }
       }
 
-    case _: TimestampNTZType => (d: String) =>
-      nullSafeDatum(d, name, nullable, options) { datum =>
-        timestampNTZFormatter.parseWithoutTimeZone(datum, false)
-      }
-
-    case _: DateType => (d: String) =>
+    case _: TimestampType => (d: String) =>
       nullSafeDatum(d, name, nullable, options) { datum =>
         try {
-          dateFormatter.parse(datum)
+          timestampFormatter.parse(datum)
         } catch {
           case NonFatal(e) =>
             // If fails to parse, then tries the way used in 2.0 and 1.x for backwards
             // compatibility.
             val str = DateTimeUtils.cleanLegacyTimestampStr(UTF8String.fromString(datum))
-            DateTimeUtils.stringToDate(str).getOrElse(throw e)
+            DateTimeUtils.stringToTimestamp(str, options.zoneId).getOrElse {
+              // There may be date type entries in timestamp column due to schema inference
+              if (options.inferDate) {
+                daysToMicros(dateFormatter.parse(datum), options.zoneId)
```

Comment:
We do not use the legacy DateFormatter here to avoid parsing timestamps with invalid suffixes. We want to throw an error. 

e.g. The legacy DateFormatter will parse the following string without throwing an error:
dateFormat: `yyyy-mm-dd`
string: `2001-09-08-randomtext`

Reply:
Yeah, I think it makes sense to throw an exception or disallow when legacy parser is used (that doesn't care about surffixes)

Reply:
>We do not use the legacy DateFormatter here to avoid parsing timestamps with invalid suffixes.

I think you could still make it work, but you would need a  new extension of `LegacySimpleDateFormatter` (maybe `LegacyStrictSimpleDateFormatter`), with an override like this:

```
  def parseToDate(s: String): Date = {
    val pp = new ParsePosition(0)
    val res = sdf.parse(s, pp)
    if (s.length != pp.getIndex) {
      throw new RuntimeException(s"$s is not a date")
    }
    res
  }
```
`2001-09-08-randomtext` would not parse, neither would `2022-01-02 12:56:33`, but `2022-01-02` would (assuming a format of `yyyy-MM-dd`).

I assume it would be slow (but I have not tested it).

Maybe not worth the extra code.

Reply:
>Do you know what is the advantage of allowing Legacy Formatter?

One benefit of the legacy formatter is that it recognizes some pre-Gregorian leap years (like `1500-02-29`) that exist only in the hybrid Julian calendar. Note how schema inference chooses `string` until you set the legacy parser.
```
scala> val csvInput = Seq("1425-03-22T00:00:00", "2022-01-01T00:00:00", "1500-02-29T00:00:00").toDS()
csvInput: org.apache.spark.sql.Dataset[String] = [value: string]

scala> spark.read.options(Map("inferSchema" -> "true", "timestampFormat" -> "yyyy-MM-dd'T'HH:mm:ss")).csv(csvInput).printSchema
root
 |-- _c0: string (nullable = true)

scala> sql("set spark.sql.legacy.timeParserPolicy=legacy")
res1: org.apache.spark.sql.DataFrame = [key: string, value: string]

scala> spark.read.options(Map("inferSchema" -> "true", "timestampFormat" -> "yyyy-MM-dd'T'HH:mm:ss")).csv(csvInput).printSchema
root
 |-- _c0: timestamp (nullable = true)

scala> 
```
That, of course, matters only if the application's input comes from legacy systems that still use hybrid Julian, and the input contains pre-Gregorian dates (e.g., for date encoding, which is the only real-world use case I have come across). I would imagine that audience is small and probably getting smaller.

Reply:
>I think you could still make it work, but you would need a new extension of LegacySimpleDateFormatter 

By the way, to avoid confusion, I meant the above in the context of inferring dates when using the legacy parser (I realize now that this discussion is happening in reference to code changes in `UnivocityParser`).

Reply:
Thanks Bruce! This is great context! This will definitely be necessary if we want to support inference along with legacy date formats. Users on legacy dates will be unaffected by this change - how about we can open another ticket for date inference with legacy formats if the demand exists (and merge this PR without legacy date inference support)?